rs/replicated_state/src/canister_state/system_state.rs

```diff
@@ -268,6 +271,232 @@ impl CanisterHistory {
     }
 }
 
+/// `TaskQueue` represents the implementation of a queue structure for tasks.
+#[derive(Clone, Eq, PartialEq, Debug, Default)]
+pub struct TaskQueue {
+    /// Queue of tasks.
+    queue: VecDeque<ExecutionTask>,
+
+    /// Status of low_on_wasm_memory hook execution.
+    on_low_wasm_memory_hook_status: OnLowWasmMemoryHookStatus,
+}
+
+impl TaskQueue {
+    pub fn from_checkpoint(
+        queue: VecDeque<ExecutionTask>,
+        on_low_wasm_memory_hook_status: OnLowWasmMemoryHookStatus,
+        canister_id: &CanisterId,
+    ) -> Self {
+        let queue = TaskQueue {
+            queue,
+            on_low_wasm_memory_hook_status,
+        };
+
+        // Because paused tasks are not allowed in checkpoint rounds when
+        // checking dts invariants that is equivalent to disabling dts.
+        queue.check_dts_invariants(
+            FlagStatus::Disabled,
+            ExecutionRoundType::CheckpointRound,
+            canister_id,
+        );
+
+        queue
+    }
+
+    pub fn front(&self) -> Option<&ExecutionTask> {
+        self.queue.front()
+    }
+
+    pub fn pop_front(&mut self) -> Option<ExecutionTask> {
+        self.queue.pop_front()
+    }
```

Comment:
How would one go about popping the `OnLowWasmMemory` task? This method or a separate one? (I'd very much prefer the former.)

Reply:
It will be in this method, just in the last PR for the feature when we enable hook.